### ISSUE TITLE
Llama 2 inference - difference compared to raw weights?

### ISSUE THREAD

**Issue by jsc1029** [Lightning-AI/lit-gpt#659]
I'm attempting to replicate the results using the raw Llama 2 weights from Meta along with the example inference scripts at https://github.com/facebookresearch/llama and getting slightly different results in the forward pass.

For example, here is the input tensor (prompt tokens) and logits from lit-gpt's base.py:

```
x: tensor([[    1,  3251,  1230,   319, 29902,   338,   263, 29871]]) 
logits: tensor([-4.9062, -7.5000,  4.9688,  ...,  2.9375, -1.1484,  0.3320])
```

Here is the same result using the raw Llama 2 weights and scripts in the original repository:

```
x: tensor([[    1,  3251,  1230,   319, 29902,   338,   263, 29871]])
logits: tensor([[-4.8359, -7.3516,  4.9922,  ...,  2.9707, -1.1104,  0.3804]])
```

Due to the slightly different logits and autoregressive nature, this tends to generate very similar text to start but eventually steers off.

**lit-gpt output (prompt: "Generative AI is a ")** 

> Generative AI is a 21st-century technology that is transforming the way we live and work. It is a type of AI that uses deep learning algorithms to generate new content, such as images, text, and music. Generative AI has the potential to revolutionize a wide range of industries, from healthcare to finance to entertainment.
> The technology behind generative AI is based on the principle of deep learning, which is a type of machine learning that uses artificial neural networks to learn from data. These networks are made up of interconnected nodes, which are trained to recognize patterns in data and make predictions. The deeper the network, the more complex the pattern recognition and the more accurate the predictions.
> One of the most promising applications of generative AI is in the field of image generation. By using deep learning algorithms, generative AI can create realistic images that are indistinguishable from those created by humans. This technology has the potential to revolutionize the way we create images for advertising, art, and other purposes.
> Another exciting application of generative AI is in the field of text generation. By using deep learning algorithms, generative AI can create realistic text that is indistinguish

**raw weights output (prompt: "Generative AI is a ")** 

> Generative AI is a 21st-century technology that is transforming the way we live and work. It is a type of AI that uses algorithms to generate new data, such as images, music, or text. It is a powerful tool that is being used in a variety of industries, including healthcare, finance, and marketing. One of the most exciting applications of generative AI is in the field of healthcare. It can be used to generate personalized treatments for patients, as well as to predict and prevent disease. For example, generative AI can be used to create personalized drug therapies based on a patient’s genetic makeup. This can help doctors to better tailor treatments to individual patients, leading to better outcomes and fewer side effects. Another exciting application of generative AI is in the field of finance. It can be used to create personalized investment strategies, as well as to predict market trends. For example, generative AI can be used to create a model that predicts the price of a stock based on its historical performance. This can help investors to make more informed decisions and to avoid

Is the most likely cause for this just a difference in the HF converted weights vs. the raw weights?  Or might it be something else?

**Comment by rasbt** [comment]
Thanks for the analysis. 

As I remember, the original weights by Meta were bfloat 16 precision whereas HF converted them to regular 16 bit precision. 

As far as I can tell though, there was no clipping going on though (the weights were all close to 0 and there was no large outlier that could have caused issues during the conversion).

So, here, theoretically the weights should be the same in 16 and bf16 bit precision. But perhaps there was some small issue or discrepancy in their conversion procedure.

Given how popular Llama 2 is, maybe we should support the original meta weights in addition to the HF weights @carmocca ?

**Comment by carmocca** [comment]
Yes! We already track this in https://github.com/Lightning-AI/lit-gpt/issues/463 so closing this one
